README.md
<p align="center">
  <a href="http://www.adjoint.io"><img src="packages/uplink.png" width="250"/></a>
</p>
<h3 align="center">Community Edition</h3>

Uplink
======

[![CircleCI](https://circleci.com/gh/adjoint-io/uplink.svg?style=svg)](https://circleci.com/gh/adjoint-io/uplink)

**Uplink** is a distributed database built for orchestrating secure multiparty
workflows found in financial markets. 

Uplink operates as a federation of nodes which communicate over a private
network to provide an append-only database in which all transactions are
cryptographically signed and where all members participate in a consensus
protocol to maintain and verify the ledger state. 

Uplink also has the capacity to run so-called *smart contracts* which are units
of logic which run across the network and allow counterparties to interact
through arbitrarily sophisticated programs that can model common business
processes and financial instruments. Provided with the ledger is a new scripting
language [FCL](src/Script) (**Financial Core Language**) built on a verified
core and designed as a target for contract modeling tools.

The community version of [Adjoint's](http://www.adjoint.io) software is released
under an Apache License and is part of a larger suite of tools and languages
offered commercially.

Running a Node
--------------

To run an Uplink node that has the ability to construct and sign blocks, you
must supply a filepath to an existing private key located in the 
`config/validators` directory. Accounts correspoding to these key pairs are 
created on boot and added to the genesis world state, and the list of addresses
defining the validator nodes is defined in `config/chain.config.local`.

#### Without Docker  

```bash
$ stack install --no-docker
```

To boot a validator node (needed to create & sign blocks): 

```bash
$ uplink chain init -b "leveldb:///uplink1" -p 8001 -d node1 -k config/validators/auth0/key -v
```

And two non-validator nodes:

```bash
$ uplink chain init -b "leveldb:///uplink2" -p 8002 --rpc-port 8546 -d node2 -v
$ uplink chain init -b "leveldb:///uplink3" -p 8003 --rpc-port 8547 -d node3 -v
```

When prompted whether we want to provide a private key, you can enter `n`.

We can open up the console:

```bash
$ uplink console
```

Running the command `listPeers` in the Uplink console should give us a
list containing the three peers we just started.

#### With Docker

The `run-container` script will compile and run a node using docker and 
bind the P2P and RPC ports to the host. 

```bash
$ stack install
```

```bash
$ ./run-container 
Please supply three args: <db dir suffix> <p2pPort> <rpcPort>
Examples: $ ./run-container 1 8001 8545
          $ ./run-container 2 8002 8546
```

To boot a validator node (needed to create & sign blocks):

```bash
$ ./run-container 1 8001 8545 "-k config/validators/auth0/key" 
```

And two non-validator nodes:

```bash
$ ./run-container 2 8002 8546
$ ./run-container 3 8003 8547
```

When prompted whether we want to provide a private key, you can enter `n`.

We can open up the console:

```bash
$ uplink console --hostname 127.0.1.1
```

Running the command `listPeers` in the Uplink console should give us a
list containing the three peers we just started.

With optimizations:

```bash
$ stack build --flag uplink:optimized
```

If you follow either of these examples, booting 3 Uplink nodes (one validator
and two non-validating nodes), then transactions issued to any of the nodes will
propagate through the network and blocks will be generated by the validating
node at intervals defined in the `blockPeriod` field of `config/chain.config`.

Uplink Console
--------------

To interact with an Uplink node with a separate console process (instead of
through the RPC), refer to the [documentation](https://www.adjoint.io/docs/console.html).

Interactive Smart Contract (FCL) Shell
--------------------------------------

```bash
$ uplink scripts 
Available options:
  -h,--help                Show this help text

  Available commands:
    compile                  Compile and typecheck a script.
    repl                     Compile, typecheck a script, and load it into a REPL.
    format                   Format a script
    lint                     Lint a script.
    graph                    Extract graph from a script.

```

```bash
$ uplink scripts repl <fcl-file>
```

Documentation
-------------

Extensive documentation can be found [here](https://www.adjoint.io/docs).


CHANGELOG.md
Changelog
========

1.7
---

*June 20, 2018*

* Add support for TLS on top of TCP transport layer
* Make transport layer configurable: TCP, TCP/TLS, or in-memory 
* Add more robust supervision and recovery of Uplink Processes
* FCL: Added Helper functions 
* FCL: Disallow global variable name shadowing
* FCL: Added generic collection values and higher-order primops operating over
  collections
* Bug fixes

1.6
---

*April 20, 2018*

* Simplify internal Node-to-Node messaging protocol by deriving the serialzation
  using GHC.Generics.
* Fix hostname resolution on all supported operating systems.
* All superfluous ByteString usage is removed.
* Added the EncodedByteString typeclass and Base16, Base58, Base64, and Base64P
  ByteString types to allow for typesafe ByteString encoding representations 
  throughout the codebase.
* Added network-access-token such that all messages exchanged by nodes in the
  Uplink network must contain proof of possession of the network-access token;
  Uplink nodes that do not possess the preshared, private network-access-token 
  will not be able to interact with an Uplink network in which membership is 
  defined by the network-access-token.
* All messages sent by Uplink nodes are now signed by the network-access-token
  private key, and all message received by Uplink nodes are verified to be from
  a node that possess the shared, private network-access-token.

1.5
---

*March 26, 2018*

* FIX: Initializing a node using `uplink chain init` with a LevelDB backend now
  (correctly) fails when a LevelDB database already exists at the supplied
  database directory.
* FIX: Exporting Ledger and Blocks as XML using a LevelDB database path now
  (correctly) requires an existing database and does not instead create an empty
  database and export nothing.
* Users can now export Uplink blocks and/or ledger state as JSON. 
* Users can now import Uplink blocks and/or ledger state when initializing an
  Uplink node with `uplink chain init`.
* Added Simulation Msg to allow users to query contract simulation ledger state.
* Addresses for Assets and Contracts are now derived by base58 encoding the hash
  of the hash of their origin transaction.
* Timestamp fields are removed from transactions due to the fact that
  transaction ordering in a block and block timestamps carry more information
  about when the transacation modified the ledger.
* Transaction uniqueness is now enforced across the entire history of the ledger
  using the sha3_256 hash of transaction datastructure contents.
* Rewrote REPL to use Simulation Process. Added option to specify exported
  ledger state to simulate the contract in. Added tab completion on methods
  callable in current state.

1.4
---

*March 13, 2018*

* Top-level declarations may now have definitions that are arbitrary
  expressions (without side-effects).
* Effect checking: we can now check and show what side-effects
  expressions or methods have.
* Fix timeout issues in Federated Consensus Algorithm. Block period is 
  now adhered to properly within the block generation process.

1.3
---

*February 26, 2018*

* Added metadata field of assets.
* Add support for enumeration types in FCL.
* References to uninitialized global variables are now rejected at 
  compile-time
* Improved graph analysis on FCL contracts.
* Fix json serialization of datetime values in contract storage
* Change FCL asset type to explicitly reference 8 distinct asset types 
* Add polymorphic asset FCL prim ops that operate over all asset types
* Added FCL Smart Contract simulation feature
* Made logging more configurable: different services can log to
  different files.
* Removed superfluous "return" construct from FCL
	
1.2
---

*January 26, 2018*

* Fix database initialization bugs on corner cases.
* Removed necessity for a Postgres database to exist with the same name of the 
  Postgres user creating the uplink database.
* Consolidated Node data file reads & writes into a single module.
* Added data query language for querying ledger state.
* Added data query interface to RPC for doing cross contract/position
  aggregations for reporting.
* Bazel build scripts for continuous integration.

1.1
---

*December 5, 2017*

* Added PostgreSQL persistence backend.
* Added Transaction "graveyard'" for transactions for reporting on transactions
  that fail to commit.
* Added database reset mechanism in test mode for provisioning test
  environments.
* Added `Circulate` transaction for assets.
* Added `RevokeAsset` transaction for assets.
* Added protocol for swappable data persistence, supporting LevelDB and SQL
  backends.
* Improved P2P performance using new asynchronous messaging handling. This will
  yield better performance when the network scales by number of nodes and
  geographical sparsity.
* Fixed issues with database and log filepath locations relative to
  configuration.
* Remove ``nodekey`` from accounts.
* Fix Mac OSX filepath bugs.

1.0
---

*October 25, 2017*

Initial release.


exe/Main.hs
{-# LANGUAGE ApplicativeDo #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE OverloadedStrings #-}

module Main (
  main,
) where

import Protolude hiding ((<>), option)

import Data.Monoid ((<>))
import Data.List (unlines, words)
import qualified Data.ByteString.Char8 as BS

import System.IO (hSetBuffering, BufferMode(..))

import Options.Applicative

-- Application logic
import Opts hiding (verbose)
import qualified Config
import qualified Driver
import qualified Account
import qualified Version
import qualified Address

title :: [Char]
title = "Uplink" <> " " <> Version.version

-- Uses evil template Haskell to get git branch at compile-time.
banner :: [Char]
banner = unlines [
    title
  , "Copyright Adjoint Inc."
  , "Branch   : " <> Version.branch
  , "Revision : " <> Version.commit
  , "Dirty    : " <> Version.dirty
  ]

-------------------------------------------------------------------------------
-- Chain
-------------------------------------------------------------------------------

-- | Parse the options from the commandline.
chainParser :: Parser Opts
chainParser = runChainParser
          <|> initChainParser
  where
    runChainParser :: Parser Opts
    runChainParser = do
      chainOpts <- optsParser
      return chainOpts { _command = Driver.Chain Run }

    initChainParser :: Parser Opts
    initChainParser =
        subparser $ command "init" $
          info (helper <*> initChainOptsParser) $
            progDesc "Initialize a new Uplink node and database."
      where
        initChainOptsParser :: Parser Opts
        initChainOptsParser = do
          init <- Init <$> initAccPromptParser <*> importDataParser
          chainOpts <- optsParser
          return chainOpts { _command = Driver.Chain init }

        initAccPromptParser :: Parser (Maybe Account.AccountKeyOpt)
        initAccPromptParser =
            -- first try
            existingKeyParser <|> newKeyParser
          where
            newKeyParser =
              flag Nothing (Just Account.New) $
                  long "key-gen"
                <> help "Generate a new key pair for the account automatically"

            existingKeyParser  =
              fmap (Just . Account.Existing) $
                strOption $
                     long "existing-key"
                  <> short 'k'
                  <> help "Create an account with an existing private key"

        -- Load blocks or world state from a file
        importDataParser :: Parser (Maybe ImportData)
        importDataParser = optional $
            importBlocksParser <|> importLedgerParser
          where
            -- Load blocks from a file
            importBlocksParser :: Parser ImportData
            importBlocksParser =
              fmap ImportBlocks $
                strOption $
                     long "import-blocks"
                  <> metavar "FILE"
                  <> help "Import blocks from a file storing blocks as XML or JSON."

            -- Load ledger state from a file
            importLedgerParser :: Parser ImportData
            importLedgerParser =
              fmap ImportLedger $
                strOption $
                     long "import-ledger"
                  <> metavar "FILE"
                  <> help "Import ledger from a file storing ledger data as JSON."

    optsParser :: Parser Opts
    optsParser = do
      _rpcPort   <- rpcPort
      _port      <- port
      _nonetwork <- nonetwork
      _chainConfig <- chainConfig
      _config    <- config
      _hostname  <- hostname
      _bootnodes <- bootnodes
      _storageBackend <- storageBackend
      _verbose   <- verbose
      _rpcReadOnly <- rpcReadOnly
      _testMode  <- testMode
      _nodeDir   <- nodeDir
      _monitorPort <- monitorPort

      pure Opts {..}


    _compile     = Nothing
    _scriptInput = Nothing
    _command     = Driver.Chain Run

keyParser :: Parser Opts
keyParser = subparser $ command "authority"
    (info (helper <*> keyParser')
    (progDesc "Manage keypairs"))
  where
    keyParser' :: Parser Opts
    keyParser' = do
      n <- nAuths
      pure defaultOpts
        { _command = Driver.Keys (Opts.CreateAuthorities n)
        }

    nAuths :: Parser Int
    nAuths = argument auto (metavar "N")

-------------------------------------------------------------------------------
-- Data Commands
-------------------------------------------------------------------------------


dataParser :: Parser Opts
dataParser = do
  cmd <- dataGetParser
     <|> dataListParser
     <|> dataExport
     <|> dataCommitParser
     <|> dataLoadAssetParser
     <|> dataLoadAccountParser
  db <- storageBackend
  pure $ cmd { _storageBackend = db }

dataGetParser :: Parser Opts
dataGetParser =  subparser $ command "get"
    (info (helper <*> dataParser')
    (progDesc "Get by address"))
  where
    dataParser' :: Parser Opts
    dataParser' = do
      addr <- addrParser
      pure defaultOpts
        { _command = Driver.Data (Get addr)
        }

dataListParser :: Parser Opts
dataListParser = subparser $ command "list"
    (info (helper <*> dataParser')
    (progDesc "List all"))
  where
    dataParser' :: Parser Opts
    dataParser' = pure defaultOpts { _command = Driver.Data List }

dataCommitParser :: Parser Opts
dataCommitParser = subparser $ command "commit"
    (info (helper <*> scriptParser')
    (progDesc "Commit local storage"))
  where
    scriptParser' :: Parser Opts
    scriptParser' = do
      compile <- fileParser
      contractAddr    <- addrParser
      accountAddr    <- addrParser
      pure defaultOpts
        { _command = Driver.Data (Commit compile contractAddr accountAddr)
        }

dataLoadAssetParser :: Parser Opts
dataLoadAssetParser = subparser $ command "loadAsset"
    (info (helper <*> scriptParser')
    (progDesc "Load Asset at Address into AssetDB"))
  where
    scriptParser' :: Parser Opts
    scriptParser' = do
      fp        <- fileParser
      assetAddr <- addrParser
      pure defaultOpts
        { _command = Driver.Data (LoadAsset fp assetAddr)
        }

dataLoadAccountParser :: Parser Opts
dataLoadAccountParser = subparser $ command "loadAccount"
    (info (helper <*> scriptParser')
    (progDesc "Load Account into AccountDB"))
  where
    scriptParser' :: Parser Opts
    scriptParser' = do
      fp        <- fileParser
      pure defaultOpts
        { _command = Driver.Data (LoadAccount fp)
        }

dataExport :: Parser Opts
dataExport =
  subparser $ command "export" $
    info (helper <*> dataExportCmdParser) $
      progDesc "Export data to file"
  where
    dataExportCmdParser :: Parser Opts
    dataExportCmdParser = do
      export <- Export
            <$> exportDataParser
            <*> fileParser
      pure defaultOpts
        { _command = Driver.Data export
        }

    exportDataParser = exportLedgerJSON <|> exportBlocks

    exportLedgerJSON :: Parser ExportData
    exportLedgerJSON =
      subparser $ command "ledger" $
        info (helper <*> pure ExportLedgerState) $
          progDesc "Export ledger data as JSON to a file."

    exportBlocks :: Parser ExportData
    exportBlocks =
      subparser $ command "blocks" $
        info (helper <*> fmap ExportBlocks formatParser) $
          progDesc "Export blocks to a file in a given format"

-------------------------------------------------------------------------------
-- Script Commands
-------------------------------------------------------------------------------

scriptParser :: Parser Opts
scriptParser =
  scriptCompileParser
  <|> scriptFormat
  <|> scriptLint
  <|> scriptGraph

scriptFormat :: Parser Opts
scriptFormat = subparser $ command "format"
    (info (helper <*> scriptParser')
    (progDesc "Format a script"))
  where
    scriptParser' :: Parser Opts
    scriptParser' = do
      compile <- fileParser
      inplace' <- inplace
      pure defaultOpts
        { _command = Driver.Script (Format compile)
        }

    inplace :: Parser Bool
    inplace = flag False True $
         long "inplace"
      <> short 'i'
      <> help "Modify file in place"


scriptCompileParser :: Parser Opts
scriptCompileParser = subparser $ command "compile"
    (info (helper <*> scriptParser')
    (progDesc "Compile and typecheck a script."))
  where
    scriptParser' :: Parser Opts
    scriptParser' = do
      compile <- fileParser
      localStorage <- localStorage
      pure defaultOpts
        { _command = Driver.Script (CompileScript compile localStorage)
        }

    localStorage :: Parser (Maybe FilePath)
    localStorage = optional $ strOption $
         long "localStorage"
      <> short 's'
      <> metavar "PATH"
      <> help "JSON to check locals against"


scriptLint :: Parser Opts
scriptLint = subparser $ command "lint"
    (info (helper <*> scriptParser')
    (progDesc "Lint a script."))
  where
    scriptParser' :: Parser Opts
    scriptParser' = do
      compile <- fileParser
      pure defaultOpts
        { _command = Driver.Script (Lint compile)
        }


scriptGraph :: Parser Opts
scriptGraph = subparser $ command "graph"
    (info (helper <*> scriptParser')
    (progDesc "Extract graph from a script."))
  where
    scriptParser' :: Parser Opts
    scriptParser' = do
      compile <- fileParser
      pure defaultOpts
        { _command = Driver.Script (Graph compile)
        }


-------------------------------------------------------------------------------
  -- Console Command
-------------------------------------------------------------------------------

consoleParser :: Parser Opts
consoleParser = consoleParser'
  where
    consoleParser' :: Parser Opts
    consoleParser' = do
      _host <- hostname
      _port <- port
      pure defaultOpts { _port = _port, _hostname = _host, _command = Console }

-------------------------------------------------------------------------------
  -- Repl Command
-------------------------------------------------------------------------------
replParser :: Parser Opts
replParser = replParser'
  where
    replParser' :: Parser Opts
    replParser' = do
      _host <- hostname
      _port <- port
      _script <- fileParser
      _verbose <- verbose

      _worldState <- optional fileParser
      pure defaultOpts { _port = _port, _hostname = _host, _command = (Repl _script (fromMaybe False _verbose) _worldState)  }


-------------------------------------------------------------------------------
-- Parser Utils
-------------------------------------------------------------------------------

fileParser :: Parser [Char]
fileParser = strArgument (metavar "FILE")

addrParser :: Parser (Address.Address a)
addrParser = argument (eitherReader addrParser') (metavar "ADDR")
  where
    addrParser' :: [Char] -> Either [Char] (Address.Address a)
    addrParser' a =
      first toS (Address.parseAddress (BS.pack a))

formatParser :: Parser DataFormat
formatParser =
  fmap (fromMaybe JSON) $
    optional $ option auto $
      mconcat [ long "format"
              , short 'f'
              , help "Format of data output"
              ]

port :: Parser (Maybe Int)
port = optional $ option auto $
     long "port"
  <> short 'p'
  <> metavar "PORT"
  <> help "Override the port on which the node should listen for P2P. Default 8001"

rpcPort :: Parser (Maybe Int)
rpcPort = optional $ option auto $
     long "rpc-port"
  <> metavar "PORT"
  <> help "Override the port on which the node should listen for RPC. Default: 8545"

config :: Parser (Maybe FilePath)
config = optional $ strOption $
     long "config"
  <> short 'c'
  <> metavar "PATH"
  <> help "Specify the node configuration file"

chainConfig :: Parser (Maybe FilePath)
chainConfig = optional $ strOption $
     long "chain-config"
  <> short 'g'
  <> metavar "PATH"
  <> help "Specify the chain (genesis block & consensus) configuration to initialize chain"

storageBackend :: Parser (Maybe [Char])
storageBackend = optional $ strOption $
     long "backend"
  <> short 'b'
  <> metavar "URI"
  <> help "Specify the storage backend uri"

genesis :: Parser (Maybe FilePath)
genesis = optional $ strOption $
     long "genesis"
  <> short 'g'
  <> metavar "PATH"
  <> help "Specify the genesis block configuration to initialize chain"

nonetwork :: Parser (Maybe Bool)
nonetwork = optional $ flag False True $
     long "no-network"
  <> help "Disable p2p networking."

hostname :: Parser (Maybe [Char])
hostname = optional $ strOption $
     long "hostname"
  <> short 'h'
  <> metavar "HOST"
  <> help "Specify hostname"

bootnodes :: Parser (Maybe [ByteString])
bootnodes = optional $ fmap (map BS.pack . words) <$> strOption $
     long "bootnodes"
  <> help "Specify boot nodes"

verbose :: Parser (Maybe Bool)
verbose = optional $ flag False True $
     long "verbose"
  <> short 'v'
  <> help "Verbose logging"

rpcReadOnly :: Parser (Maybe Bool)
rpcReadOnly = optional $ flag False True $
     long "rpc-read-only"
  <> help "RPC read only"

testMode :: Parser (Maybe Bool)
testMode = optional $ flag False True $
     long "test"
  <> help "Run node in test mode"

nodeDir :: Parser (Maybe FilePath)
nodeDir = optional $ strOption $
     long "datadir"
  <> short 'd'
  <> metavar "PATH"
  <> help "Specify the node data directory path"

monitorPort :: Parser (Maybe Int)
monitorPort = optional $ option auto $
     long "monitor-port"
  <> metavar "PORT"
  <> help "Specify the port the monitoring server runs on"

-------------------------------------------------------------------------------
-- Toplevel
-------------------------------------------------------------------------------

versionCmd :: Parser Opts
versionCmd = pure defaultOpts { _command = Version }

copts :: Parser Opts
copts = subparser
  (
      command "chain"    (info (helper <*> chainParser)  (progDesc "Manage and setup blockchains."))
   <> command "keys"     (info (helper <*> keyParser)    (progDesc "Manage keypairs."))
   <> command "scripts"  (info (helper <*> scriptParser) (progDesc "Manage smart contract scripts."))
   <> command "data"     (info (helper <*> dataParser)   (progDesc "Manage ledger data and oracle services."))
   <> command "console"  (info (helper <*> consoleParser)(progDesc "Connect to a node and issue transactions."))
   <> command "repl"     (info (helper <*> replParser)   (progDesc "Connect to a node and simulate a contract."))
   <> command "version"  (info (helper <*> versionCmd)   (progDesc "Version info"))
  )

handleOpts :: Opts -> IO ()
handleOpts o@Opts {..} = do
  hSetBuffering stdout NoBuffering

  let
    silent = case Opts._command o of
      Chain _    -> False
      Keys _     -> True
      Script _   -> True
      Data _     -> True
      Console    -> False
      Repl _ _ _ -> False
      Version    -> False

  case Opts._command o of
      Chain _    -> putStrLn banner
      Keys _     -> pure ()
      Script _   -> pure ()
      Data _     -> pure ()
      Console    -> putStrLn banner
      Repl _ _ _ -> putStrLn banner
      Version    -> putStrLn banner

  -- Load Config from command line fallbacking to 'defaultConfig'
  c@Config.Config {}
    <- Config.handleConfig silent (maybe Config.defaultConfig identity _config)

  backend <- mapM Config.getStorageBackend _storageBackend
  -- Load config falling back to defaults.
  let config @ Config.Config {..}
        = Config.toggleVerbosity _verbose
          $ c
            { Config.configFile
                = _config      `fallback` Config.defaultConfig
            , Config.nonetwork
              = _nonetwork   `fallback` (Config.nonetwork c)
            , Config.port
              = _port        `fallback` (Config.port c)
            , Config.rpcPort
              = _rpcPort     `fallback` (Config.rpcPort c)
            , Config.chainConfigFile
              = _chainConfig   `fallback` (Config.chainConfigFile c)
            , Config.hostname
              = _hostname    `fallback` (Config.hostname c)
            , Config.bootnodes
              = fromMaybe [] _bootnodes ++ Config.bootnodes c
            , Config.storageBackend
              = backend   `fallback` (Config.storageBackend c)
            , Config.rpcReadOnly
              = _rpcReadOnly `fallback` (Config.rpcReadOnly c)
            , Config.testMode
              = _testMode    `fallback` (Config.testMode c)
            , Config.nodeDataDir
              = _nodeDir     `fallback` (Config.nodeDataDir c)
            , Config.monitorPort
              = _monitorPort `fallback` (Config.monitorPort c)
            }

  Driver.driver (Opts._command o) config

  where
    fallback :: Maybe a -> a -> a
    fallback = flip fromMaybe

-- Command options
coptions :: ParserInfo Opts
coptions = info (helper <*> copts) idm

-------------------------------------------------------------------------------
-- Entry
-------------------------------------------------------------------------------

cprefs :: ParserPrefs
cprefs = prefs showHelpOnError

-- Parse commmand line flags and handle application
main :: IO ()
main = customExecParser cprefs coptions >>= handleOpts
